File: 102.binary_tree_level_order.py
"""
Given the root of a binary tree, return the level order traversal of its nodes'
values. (i.e., from left to right, level by level).
"""
"""
Solved both in about 20 minutes, starting with the iterative solution.  At first I thought it
might be possible to do the iterative solution without pushing the auxiliary depth information
but I didn't figure out how to avoid that.

How to do this recursively?  You really can't do BFS recursively.  But you could do
a DFS while keeping track of depth

"""
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        results = []
        
        # add all of the nodes in the node subtree, where
        # node is at depth
        def add_rec(node, depth):
            if not node:
                return
            if len(results) <= depth:
                results.append([])
            results[depth].append(node.val)
            add_rec(node.left, depth+1)
            add_rec(node.right, depth+1)
        
        add_rec(root, 0)
        return results
    


"""
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        from collections import deque
        d = deque([(root, 0)])
        results = []
        while d:
            node, level = d.popleft()
            if not node:
                continue
            if len(results) < level + 1:
                   results.append([])
            results[level].append(node.val)
            d.extend([(node.left, level+1), (node.right, level+1)])
        
        return results
""" 

File: 102.binary_tree_level_order.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
      vector<vector<int>> result;
      if (!root) return result;
      queue<pair<int, TreeNode*>> q;
      q.emplace(0, root);
      while (!q.empty()) {
        auto [depth, node] = q.front();
        q.pop();
        if (depth == result.size()) result.push_back(vector<int>());
        result[depth].push_back(node->val);
        if (node->left) q.emplace(depth+1, node->left);
        if (node->right) q.emplace(depth+1, node->right);
      }
      return result;
    }
};

